controllers/imageclusterinstall_controller.go

```diff
@@ -152,136 +154,230 @@ func (r *ImageClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
 		log.Infof("Disable automated cleaning mode for BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
 		if err := r.Patch(ctx, bmh, patch); err != nil {
-			return ctrl.Result{}, err
+			cond.Reason = v1alpha1.ConfigurationFailedReason
+			cond.Message = fmt.Sprintf("failed to disable automated cleaning mode for BareMetalHost %s/%s", bmh.Namespace, bmh.Name)
+			log.WithError(err).Error(cond.Message)
+			return nil, nil, err
 		}
 	}
 
-	if err = r.validateBMH(ctx, log, ici, bmh); err != nil {
-		// in case of failure to validate the BMH just requeue
-		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	return cd, bmh, nil
+}
+
+func (r *ImageClusterInstallReconciler) validateHost(
+	ctx context.Context,
+	ici *v1alpha1.ImageClusterInstall,
+	bmh *bmh_v1alpha1.BareMetalHost,
+	cd *hivev1.ClusterDeployment,
+	cond *hivev1.ClusterInstallCondition,
+	log logrus.FieldLogger,
+) (ctrl.Result, error) {
+	cond.Reason = v1alpha1.HostValidationFailedReason
+
+	if res, err := r.validateBMH(ici, bmh, cond); !res.IsZero() || err != nil {
+		return res, err
 	}
 
-	if err = r.setClusterInstallMetadata(ctx, log, ici, clusterDeployment); err != nil {
-		log.WithError(err).Error("failed to set ImageClusterInstall data")
+	if !bmh.Spec.ExternallyProvisioned {
+		log.Infof("Setting BareMetalHost (%s/%s) ExternallyProvisioned spec", bmh.Namespace, bmh.Name)
+		patch := client.MergeFrom(bmh.DeepCopy())
+		bmh.Spec.ExternallyProvisioned = true
+		if err := r.Patch(ctx, bmh, patch); err != nil {
+			return ctrl.Result{}, err
+		}
+	
+	}
+
+	if err := r.setClusterInstallMetadata(ctx, log, ici, cd); err != nil {
+		cond.Message = "failed to set ImageClusterInstall data"
+		log.WithError(err).Error(cond.Message)
 		return ctrl.Result{}, err
 	}
 
-	res, err := r.writeInputData(ctx, log, ici, clusterDeployment, bmh)
+	return ctrl.Result{}, nil
+}
+
+func (r *ImageClusterInstallReconciler) createImage(
+	ctx context.Context,
+	ici *v1alpha1.ImageClusterInstall,
+	req ctrl.Request,
+	bmh *bmh_v1alpha1.BareMetalHost,
+	cd *hivev1.ClusterDeployment,
+	cond *hivev1.ClusterInstallCondition,
+	log logrus.FieldLogger,
+) (string, ctrl.Result, error) {
+	cond.Reason = v1alpha1.ImageCreationFailedReason
+
+	res, err := r.writeInputData(ctx, log, ici, cd, bmh)
 	if !res.IsZero() || err != nil {
+		cond.Reason = v1alpha1.ImageCreationPendingReason
+		cond.Message = "could not acquire lock for image data"
 		if err != nil {
-			if updateErr := r.setImageReadyCondition(ctx, ici, err); updateErr != nil {
-				log.WithError(updateErr).Error("failed to update ImageClusterInstall status")
-			}
-			log.Error(err)
+			cond.Reason = v1alpha1.ImageCreationFailedReason
+			cond.Message = "failed to create image"
+			log.WithError(err).Error(cond.Message)
 		}
-		return res, err
+		return "", res, err
 	}
 
-	r.labelReferencedObjectsForBackup(ctx, log, ici, clusterDeployment)
+	r.labelReferencedObjectsForBackup(ctx, log, ici, cd)
```

Comment:
This isn't really part of the image creation, can this maybe just live in the caller since it's never going to cause a condition failure?